growth_rate.js

```diff
@@ -0,0 +1,117 @@
+
+// A module that shows the base of the best-fit exponential curve of the data in
+// each state.
+//
+// Currently, for each day, we take all the data up to that day, and fit an
+// exponential curve to it, and calculate the error of that fit.  We use the base
+// color of the state to show the exponent of the best-fit curve (i.e., the best
+// growth rate), and the dot to show the error.
+const GrowthRateModule = {
+	// The name under which this variable shows up in the variable selector
+	variableName: "Growth Rate",
+
+	// A function to add our data to the base data variable
+	addData: function(data, dates, states) {
+		const expErr = function (series, a, b) {
+			return series.map(entry => {
+				const [x, y] = entry
+				const expectedY = a * Math.exp(b * x)
+				return (y - expectedY) * (y - expectedY)
+			}).reduce((a, b) => a + b, 0) / series.length
+		}
+		const expFit = function (pts) {
+			const goodPts = pts.filter(pt => pt[1] > 0)
+			const n = goodPts.length
+
+			if (n > 2) {
+				const sumX = goodPts.map(pt => pt[0]).reduce((a, b) => a + b, 0)
+				const sumX2 = goodPts.map(pt => pt[0] * pt[0]).reduce((a, b) => a + b, 0)
+				const sumY = goodPts.map(pt => Math.log(pt[1])).reduce((a, b) => a + b, 0)
+				const sumXY = goodPts.map(pt => pt[0] * Math.log(pt[1])).reduce((a, b) => a + b, 0)
+				const a = Math.exp((sumX2 * sumY - sumX * sumXY) / (n * sumX2 - sumX * sumX))
+				const b = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX)
+				return [a, b]
+			} else {
+				return [0.0, 0.0]
+			}
+		}
+		for (s=0; s<states.length; s++) {
+			const state = states[s]
+			const stateData = data
+			    .filter(datum => datum.state == state)
+			      .sort((a, b) => a.doy - b.doy)
+			const pointData = stateData.map(datum => [datum.doy, datum.positive])
+			const minDay = pointData.map(datum => datum[0]).reduce((a, b) => Math.min(a, b))
+			const maxDay = pointData.map(datum => datum[0]).reduce((a, b) => Math.max(a, b))
+			for (d = minDay; d <= maxDay; ++d) {
+				const dailySlice = pointData.slice(0, d - minDay)
+				const dailyFit = expFit(dailySlice)
+				if (dailyFit[1] > 0) {
+					const dailyErr = expErr(dailySlice,
+											dailyFit[0], dailyFit[1])
+					stateData[d - minDay].exp = dailyFit[1]
+					stateData[d - minDay].expErr = dailyErr
+				}
+			}
+		}
+	},
+
+	// A function to copy data from the root data structure to the vals structure used for display
+	copyData: function (root, display) {
+		if (root.exp) {
+			display.exp = Math.exp(root.exp)
+			display.expErr = root.expErr
+		} else {
+			display.exp = 0.0
+			display.expErr = 0.0
+		}
+	},
+	
+	// A function that gives the value for a given feature
+	valueFcn: function (feat, date) {
+		return getValue(feat, date, 'exp', false)
+	},
+
+	// A function that gives the text to use for a given feature
+	valueTextFcn: function (feat, date) {
+		return "Growth rate: " + getValue(feat, date, 'exp', false)
+	},
+
+	// The color scale to use for our main variable
```

Comment:
legendColorScale should come before the (choropleth) color scale, so that the legendColorScale can be used inside the choropleth color scale. The only difference is that zero is omitted from the legend. In fact we should merge these, but I'm not sure how yet...

More importantly, the colorScale function should return a value (a color), not a function. This is why you are not getting any colors.

Finally, what is the range of values you expect? I am getting values in the hundreds, which means your domain of [1,3] doesn't cover the values.

Reply:
I think the issues with the numbers and colors in the scale are all now fixed - though the labels aren't what I want them to be still.